server/consumer.go

```diff
@@ -3539,6 +3551,68 @@ var (
 	errWaitQueueNil  = errors.New("wait queue is nil")
 )
 
+// insertSorted inserts wr at the correct position based on priority
+func (wq *waitQueue) insertSorted(wr *waitingRequest) {
+	priority := math.MaxInt32
+	if wr.priorityGroup != nil {
+		priority = wr.priorityGroup.Priority
+	}
+
+	// Handle empty queue
+	if wq.head == nil {
+		wq.head = wr
+		wq.tail = wr
+		wr.next = nil
+		return
+	}
+
+	// Find insertion point - insert before first element with higher priority
+	var prev *waitingRequest
+	current := wq.head
+
+	for current != nil {
+		currentPriority := math.MaxInt32
+		if current.priorityGroup != nil {
+			currentPriority = current.priorityGroup.Priority
+		}
+		// Insert before the first element with higher priority (stable sort)
+		if currentPriority > priority {
+			break
+		}
+		prev = current
+		current = current.next
+	}
+
+	// Insert at found position
+	if prev == nil {
+		// Insert at head (lowest priority so far)
+		wr.next = wq.head
+		wq.head = wr
+	} else {
+		// Insert after prev
+		wr.next = prev.next
+		prev.next = wr
+		if wr.next == nil {
+			// We're the new tail
+			wq.tail = wr
+		}
+	}
```

Comment:
This same logic is also used in `popAndRequeue`. Could perhaps extract this logic and reuse it in both `insertSorted` and `popAndRequeue`.